### PR TITLE
PMS-12 restock and sell

### PR DESCRIPTION
**Pull request by julianneviz**
Screenshot:
image
image

Implemented:
- restock item (html, ccs, js)
- sell item (html, ccs, js)
- restock and sell item updates quantity at db
- asynchronous loading to the table displayed
- error message displayed for restock : invalid code
- error messages displayed for sell: invalid code, insufficient stock, no available stock

Need to watch out for:
- unsure on how error messages are displayed

### CODE REVIEW

File: controllers/itemController.js [hunk 1]
```diff
@@ -83,6 +83,26 @@ const itemController = {
 
     },
 
+    checkItem: async function (req, res) {
```
Comment (aar0npal) [inline controllers/itemController.js]: Whats the difference between this and getItem function?

File: public/js/index.js [hunk 2]
```diff
@@ -70,7 +70,7 @@ function item(
 }
 
 // On document ready
-$(function () {
+$(document).ready(function(){
```
Comment (aar0npal) [inline public/js/index.js]: Okay lang si  `$(function () {`

File: public/js/index.js [hunk 3]
```diff
@@ -159,34 +159,173 @@ $(function () {
...
+
+        const data = new FormData($("#restock-form")[0]);
+        const code = $("#restock-popup #code").val();
+        var recID = w2ui["itemGrid"].find({ code:  code});
+        recID = recID[0];
+
+        $.ajax({
+            url: "/checkItem",
+            data: data,
+            type: "POST",
```
Comment (aar0npal) [inline public/js/index.js]: Use `GET`

File: public/js/index.js [hunk 4]
```diff
@@ -159,34 +159,173 @@ $(function () {
...
+    $("#restock-popup form .command :submit").on("click", function(e) {
+        e.preventDefault();
+
+        const data = new FormData($("#restock-form")[0]);
+        const code = $("#restock-popup #code").val();
+        var recID = w2ui["itemGrid"].find({ code:  code});
+        recID = recID[0];
+
+        $.ajax({
+            url: "/checkItem",
```
Comment (aar0npal) [inline public/js/index.js]: Use `/getItem/${code}`. You can use string interpolation.

File: routes/routes.js [hunk 5]
```diff
@@ -10,8 +10,12 @@ const app = express();
 // The dashboard or inventory page
 app.get("/", checkAuth, itemController.home);
 app.post("/addItem", upload.single("image"), itemController.addItem);
+app.post("/checkItem", upload.any(), itemController.checkItem);
```
Comment (aar0npal) [inline routes/routes.js]: Omit `app.post('/checkItem'...)`.
Use `app.get('/getItem/:code'...)` then use `req.params.code` to get the item code.

File: public/js/index.js [hunk 6]
```diff
@@ -159,34 +159,173 @@ $(function () {
...
+            data: data,
```
Comment (aar0npal) [inline public/js/index.js]: `data` is not required for GET http requests.

File: public/js/index.js [hunk 7]
```diff
@@ -159,34 +159,173 @@ $(function () {
...
+        e.preventDefault();
+
+        const quantity = $('#sell-popup #quantity').val();
+        const data = new FormData($("#sell-form")[0]);
+        const code = $("#sell-popup #code").val();
+        var recID = w2ui["itemGrid"].find({ code:  code});
+        recID = recID[0];
+
+        $.ajax({
+            url: "/checkItem",
```
Comment (aar0npal) [inline public/js/index.js]: Same here.

File: public/js/index.js [hunk 8]
```diff
@@ -159,34 +159,173 @@ $(function () {
...
+                    }
+                    else {
+                        $.ajax({
+                            url: "/sellItem",
+                            data: data,
+                            type: "POST",
+                            processData: false,
+                            contentType: false,
+
+                            success: async function (flag) {
```
Comment (aar0npal) [inline public/js/index.js]: Move the error checking under `error` of ajax.

**Review by aar0npal — CHANGES_REQUESTED** [review]

**Review by aar0npal — APPROVED** [review]

File: controllers/itemController.js [hunk 9]
```diff
@@ -9,7 +9,7 @@ const itemController = {
     home: function (req, res) {
         res.render("index", {
             title: "index",
-            styles: ["index.css", "w2ui-overrides.css", "popup.css"],
+            styles: ["index.css", "w2ui-overrides.css", "./general/popup.css"],
```
Comment (aar0npal) [inline controllers/itemController.js]: Remove `./`

File: controllers/itemController.js [hunk 10]
```diff
@@ -76,11 +76,47 @@ const itemController = {
...
+            if (data){
+                res.status(200).json(await data);
+            }
+            else {
+                res.status(400).json({message: "Invalid Product Code.", fields: ["code"]});
+            }
+        });
+    },
+
+    checkItem: async function (req, res) {
```
Comment (aar0npal) [inline controllers/itemController.js]: Remove `checkItem`

File: public/js/index.js [hunk 11]
```diff
@@ -159,48 +159,227 @@ $(function () {
...
+                $.ajax({
+                    url: "/restockItem",
+                    data: data,
+                    type: "POST",
+                    processData: false,
+                    contentType: false,
+
+                    success: async function (foundData) {
+                        $.ajax({
+                                url: `/getItem/${code}`,
```
Comment (aar0npal) [inline public/js/index.js]: Use `/getItem=:code`

File: routes/routes.js [hunk 12]
```diff
@@ -20,7 +23,7 @@ app.delete("/auth/logout", userController.logoutUser);
 
 // The Item Page
 app.get("/item/:code", checkAuth, itemController.itemDetails);
-app.get("/getItem", itemController.getItem);
+app.get("/getItem/:code", itemController.getItem);
```
Comment (aar0npal) [inline routes/routes.js]: Change to `/getItem=:code`

**Review by aar0npal — CHANGES_REQUESTED** [review]

**Review by aar0npal — APPROVED** [review]
Looks good

**Comment by SeanUmpad** [comment]
**Checked:**
Clicking the ‘Restock’ and ‘Sell’ button shows a pop-up form - Working
‘Sell’ or ‘Restock’ button - Other than the missing validations it is working
‘Cancel’ button - Working

**Test Cases:** https://docs.google.com/spreadsheets/d/1-uWENiTCqkX6zCxiP422kY14om7W3PoVnO5XfAQxOH8

**Errors:** 
Test ID 021: Lacking validation for selling 0 items (might show up sa transactions kasi na sold 0 items)
Test ID 022: Lacking validation for non-numeric text in quantity when selling e.g. "testing"
Test ID 025: Lacking validation for restocking 0 items (might show up sa transactions kasi na restocked 0 items)
Test ID 026: Lacking validation for non-numeric text in quantity when restocking e.g. "testing222e"

Test ID 028-029: Lacking validation for decimals in quantity when restocking/selling (should be whole number)
image
image

Test ID 030-031 (continuing from test 28-29): If current stock quantity is a decimal the error that should be present for decimals is shown when doing a valid transaction
image

Test ID 032: Pop up form gets messed up when zooming in and out the browser
image

**Notes and Questions:**
- Should the product code be case insensitive?

**Comment by aar0npal** [comment]
> Should the product code be case-insensitive?

Yes. Take note, @julianneviz .
You can convert to uppercase when getting the code sa front-end js before passing to ajax.

**Comment by SeanUmpad** [comment]
Test ID 032: Pop up form gets messed up when zooming in and out the browser - to be fixed in other pms

**Review by SeanUmpad — APPROVED** [review]

**Comment by aar0npal** [comment]
> Test ID 032: Pop up form gets messed up when zooming in and out the browser - to be fixed in other pms

@SeanUmpad 
[T-032 Resolved](https://github.com/erik-lance/TitafindsXG5/pull/10/commits/805b24c71f8f7d9292544985402466a70b83beef)

**Pull request merged**